pusher-platform-core/src/main/kotlin/com/pusher/platform/ErrorResolver.kt

```diff
@@ -1,67 +1,46 @@
 package com.pusher.platform
 
-import com.pusher.platform.network.ConnectivityHelper
-import com.pusher.platform.retrying.DoNotRetry
-import com.pusher.platform.retrying.Retry
+import com.pusher.platform.network.Futures
+import com.pusher.platform.network.cancel
+import com.pusher.platform.retrying.RetryStrategy
+import com.pusher.platform.retrying.RetryStrategy.DoNotRetry
+import com.pusher.platform.retrying.RetryStrategy.Retry
 import com.pusher.platform.retrying.RetryStrategyOptions
-import com.pusher.platform.retrying.RetryStrategyResult
 import elements.Error
 import elements.ErrorResponse
 import elements.NetworkError
-import elements.retryAfter
 import java.util.*
+import java.util.concurrent.Future
 import kotlin.math.min
-
-typealias RetryStrategyResultCallback = (RetryStrategyResult) -> Unit
-
-fun String.isSafeRequest(): Boolean = when(this.toUpperCase()){
-    "GET", "SUBSCRIBE", "HEAD", "PUT" -> true
-    else -> false
-}
-
-fun ErrorResponse.isRetryable(): Boolean =
-    this.statusCode in 500..599 &&
-            this.headers["Request-Method"]?.firstOrNull()?.isSafeRequest() ?: false
-
+import kotlin.math.pow
 
 class ErrorResolver(
-    private val connectivityHelper: ConnectivityHelper,
     private val retryOptions: RetryStrategyOptions,
-    private val scheduler: Scheduler,
     private val retryUnsafeRequests: Boolean = false
 ) {
 
+    private val runningJobs = LinkedList<Future<*>>()
```

Comment:
Does this list grow unbounded? Or do all uses of the `ErrorResolver` eventually end with a call to `cancel()`?

Reply:
`ErrorResolver` is bound entirely to a single subscription and the cancel is called when the subscription is unsubscribed. which reminds me that this class can be `internal` and that we should check if the job is done and/or cancelled. Adding these two.

Ideally in the future, if we merge the two (redundant) types of subscriptions this could be part of the subscription itself. @hamchapman agrees that we should merge those two types...